contextualized/dags/torch_notmad/torch_notmad.py

```diff
@@ -0,0 +1,286 @@
+        super(NGAM, self).__init__()
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        
+        nam_input_layer = nn.Linear(1, width)
+        init_hidden_layer = nn.Linear(width, width)
+        nam_output_layer = nn.Linear(width, output_dim)
+        
+        with torch.no_grad(): #use uniform randomization
+            init_hidden_layer.weight = self._get_new_weights(init_hidden_layer)
+            nam_input_layer.weight = self._get_new_weights(nam_input_layer)
+            nam_output_layer.weight = self._get_new_weights(nam_output_layer)
+        
+        hidden_layers = lambda: [layer for layer in (init_hidden_layer, activation()) for _ in range(n_hidden_layers)]
+        nam_layers = lambda: [nam_input_layer, activation()] + hidden_layers() + [nam_output_layer]
+        
+        self.nams =  nn.ModuleList([nn.Sequential(*nam_layers()) for _ in range(self.input_dim)])
+        self.link_fn = link_fn
+
+    def forward(self, x):
+        batch_size = x.shape[0]
+        ret = torch.zeros((batch_size, self.output_dim))
+        for i, nam in enumerate(self.nams):
+            ret += nam(x[:, i].unsqueeze(-1)) #batch at contextual feature i
+        
+        return self.link_fn(ret)
+    
+    def _get_new_weights(self, layer_):
+        with torch.no_grad():
+            return nn.parameter.Parameter(
+                    torch.tensor(np.random.uniform(-0.01, 0.01, size=layer_.weight.shape)).float()
+                )
+    
+
+class Explainer(nn.Module):
+    """
+    Explainer module for 2D archetypes
+    """
+    def __init__(self, archetypes_shape, init_mat=None):
+        """ Initialize the Explainer.
+
+        Args:
+            archetypes_shape (int, int): Shape of archetypes (n_archetypes, # features)
+            init_mat (np.array): 3D Custom initial weights for each archetype. Defaults to None.
+        """
+        super(Explainer, self).__init__()
+        self.k, self.d = archetypes_shape
+
+        if init_mat is not None:
+            self.init_mat = torch.tensor(init_mat)
+        else:
+            self.init_mat = torch.tensor(np.random.uniform(-0.01, 0.01, size=(self.k, self.d, self.d))).float()
+        
+        self.archs = nn.parameter.Parameter(self.init_mat, requires_grad=True)
+        self.mask = torch.ones(self.d).float() - torch.eye(self.d).float()
+    
+    def forward(self,batch_weights):
+        batch_size = batch_weights.shape[0]
+        archetypes = self._archetypes()
+        
+        batch_archetypes = torch.tensor(np.zeros((batch_size, self.d, self.d)))
+        for i,batch_w in enumerate(batch_weights):
+            batch_archetypes[i] = torch.tensordot(batch_w.float(), archetypes.float(), dims=1)
+        
+        return batch_archetypes
+                
+    
+    def _archetypes(self):
+        return torch.multiply(self.archs, self.mask)
+    
+
+class NOTMAD_model(pl.LightningModule):
+    """
+    NOTMAD model
+    """
+    def __init__(self, 
```

Comment:
Try to make this consistent with the contextualized.regression.lightning_modules args e.g. opt_lr --> learning_rate, encoder_width + n_encoder_hidden_layers --> encoder_kwargs